include/Lv2Ports.h

```diff
@@ -0,0 +1,237 @@
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LV2PORTS_H
+#define LV2PORTS_H
+
+#include "lmmsconfig.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include <lilv/lilv.h>
+#include <memory>
+#include <vector>
+
+#include "lmms_basics.h"
+#include "PluginIssue.h"
+
+struct ConnectPortVisitor;
+
+namespace Lv2Ports {
+
+/*
+	port structs
+*/
+enum class Flow {
+	Unknown,
+	Input,
+	Output
+};
+
+enum class Type {
+	Unknown,
+	Control,
+	Audio,
+	Event, //!< TODO: unused, describe
+	Cv //!< TODO: unused, describe
+};
+
+//! Port visualization
+//! @note All Lv2 audio ports are float, this is only the visualisation
+enum class Vis {
+	None,
+	Integer,
+	Enumeration,
+	Toggled
+};
+
+const char* toStr(Lv2Ports::Flow pf);
+const char* toStr(Lv2Ports::Type pt);
+const char* toStr(Lv2Ports::Vis pv);
+
+struct ControlPortBase;
+struct Control;
+struct Audio;
+struct Cv;
+struct Unknown;
+
+struct ConstVisitor
+{
+	virtual void visit(const Lv2Ports::ControlPortBase& ) {}
+	virtual void visit(const Lv2Ports::Control& ) {}
+	virtual void visit(const Lv2Ports::Audio& ) {}
+	virtual void visit(const Lv2Ports::Cv& ) {}
+	virtual void visit(const Lv2Ports::Unknown& ) {}
+
+	virtual ~ConstVisitor();
+};
+
+struct Visitor
+{
+	virtual void visit(Lv2Ports::ControlPortBase& ) {}
+	virtual void visit(Lv2Ports::Control& ) {}
+	virtual void visit(Lv2Ports::Audio& ) {}
+	virtual void visit(Lv2Ports::Cv& ) {}
+	virtual void visit(Lv2Ports::Unknown& ) {}
+
+	virtual ~Visitor();
+};
+
+struct Meta
+{
+	Type m_type = Type::Unknown;
+	Flow m_flow = Flow::Unknown;
+	Vis m_vis = Vis::None;
+
+	float m_def = .0f, m_min = .0f, m_max = .0f;
+	bool m_optional = false;
+	bool m_used = true;
+
+	std::vector<PluginIssue> get(const LilvPlugin* plugin, std::size_t portNum);
+};
+
+struct PortBase : public Meta
+{
+	const LilvPort* m_port = nullptr;
+	const LilvPlugin* m_plugin = nullptr;
+
+	virtual void accept(Visitor& v) = 0;
+	virtual void accept(ConstVisitor& v) const = 0;
+
+	QString name() const;
+	QString uri() const;
+
+	virtual ~PortBase();
+};
+
+#define IS_PORT_TYPE \
+void accept(Visitor& v) override { v.visit(*this); } \
+void accept(ConstVisitor& v) const override { v.visit(*this); }
```

Comment:
It's a good idea, though we have `Control` and `Cv` both inheriting from `ControlPortBase`. In this case, it's probably best to let `ControlPortBase` inherit from `VisitablePort<ControlPortBase>` and add the two `accept` functions into `Control` and `Cv` manually?

Reply:
I guess I didn't really properly take in that the hierarchy had several layers when I made this suggestion. If you add the functions into some classes manually, you sort of lose the advantage of this technique and might as well stick with the macro solution.

It should be possible to adapt the template approach though by using a kind of CRTP-mixin hybrid:
```cpp
template<typename Derived, typename Base>
struct VisitablePort : public Base
{
	void accept(Visitor& v) override { v.visit(static_cast<Derived&>(*this)); }
	void accept(ConstVisitor& v) const override { v.visit(static_cast<const Derived&>(*this)); }
};

struct ControlPortBase : public VisitablePort<ControlPortBase, PortBase>
// or just `struct ControlPortBase : public PortBase` - see comment below
{
	// ...
};

struct Control : public VisitablePort<Control, ControlPortBase>
{
	// ...
};
```
Does `ControlPortBase` need to be able to accept visitors itself, if it's only ever used as a base? If not, it can derive directly from `PortBase` (as it does currently), and the `VisitablePort` struct template can be used just for the leaf classes.

You may think these templates are getting unnecessarily ugly and just want to stick with the macro - I'll leave that to you (although I have a bias against macros).

(A third argument might be that since each port needs adding to the visitor structs manually, the visitors should be added to each port manually. How much is it worth it to automate only half the job?)

Reply:
The CRTP solutions works. I'm not sure if it's more readable than the macros :smile: But I don't like macros either.

It's one of those topics where I can't decide what's better, but CRTP works now, we can still change it to macros if anyone should complain about it.